File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {"(":")", "{":"}", "[":"]"}
        for char in s:
            if char in mapping.keys():
                stack.append(char)

            elif not stack:
                return False

            elif char in mapping.values():
                if char != mapping[stack.pop()]:
                    return False

        return not stack

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        Map<Character, Character> mapping = Map.of('(', ')', '{', '}', '[', ']');

        for (char charInS : s.toCharArray()) {
            if (mapping.containsKey(charInS)) {
                stack.push(charInS);
            } else if (stack.isEmpty()) {
                return false;
            } else if (mapping.containsValue(charInS)) {
                if (charInS != mapping.get(stack.pop())) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }
}